Flaky-test: ManagedCursorTest.testForceCursorRecovery

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/11232853373/job/31225793977?pr=23414#step:11:2927

### Exception stacktrace

<!-- copy-paste the stack trace in the code block below -->
```
  Error:  Tests run: 282, Failures: 1, Errors: 0, Skipped: 202, Time elapsed: 36.683 s <<< FAILURE! - in org.apache.bookkeeper.mledger.impl.ManagedCursorTest
  Error:  org.apache.bookkeeper.mledger.impl.ManagedCursorTest.testForceCursorRecovery  Time elapsed: 0.031 s  <<< FAILURE!
  java.lang.AssertionError: expected [true] but found [false]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1577)
  	at org.testng.Assert.assertTrue(Assert.java:56)
  	at org.testng.Assert.assertTrue(Assert.java:66)
  	at org.apache.bookkeeper.mledger.impl.ManagedCursorTest.testForceCursorRecovery(ManagedCursorTest.java:4831)
  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  	at java.base/java.lang.Thread.run(Thread.java:1583)
```


### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
@@ -4826,9 +4826,8 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
         assertEquals(cursor.getReadPosition(), markDeletedPosition.getNext());
     }
 
-    @Test
+    @Test(invocationCount = 5000)
     void testForceCursorRecovery() throws Exception {
-        ManagedLedgerFactoryConfig managedLedgerFactoryConfig = new ManagedLedgerFactoryConfig();
         TestPulsarMockBookKeeper bk = new TestPulsarMockBookKeeper(executor);
         factory = new ManagedLedgerFactoryImpl(metadataStore, bk);
         ManagedLedgerConfig config = new ManagedLedgerConfig();
@@ -4841,17 +4840,24 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
         ManagedCursorInfo info = ManagedCursorInfo.newBuilder().setCursorsLedgerId(invalidLedger).build();
         CountDownLatch latch = new CountDownLatch(1);
         MutableBoolean recovered = new MutableBoolean(false);
+        AtomicBoolean callbackInvoked = new AtomicBoolean(false);
         VoidCallback callback = new VoidCallback() {
             @Override
             public void operationComplete() {
-                recovered.setValue(true);
-                latch.countDown();
+                if (callbackInvoked.compareAndSet(false, true)) {
+                    log.info("Cursor recovery success");
+                    recovered.setValue(true);
+                    latch.countDown();
+                }
             }
 
             @Override
             public void operationFailed(ManagedLedgerException exception) {
-                recovered.setValue(false);
-                latch.countDown();
+                if (callbackInvoked.compareAndSet(false, true)) {
+                    log.error("Cursor recovery failed", exception);
+                    recovered.setValue(false);
+                    latch.countDown();
+                }
             }
         };
         c1.recoverFromLedger(info, callback);

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
@@ -4880,6 +4880,7 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
                 final OpenCallback cb, final Object ctx) {
             if (ledgerErrors.containsKey(lId)) {
                 cb.openComplete(ledgerErrors.get(lId), null, ctx);
+                return;
             }
             super.asyncOpenLedger(lId, digestType, passwd, cb, ctx);
         }

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
@@ -4826,7 +4826,7 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
         assertEquals(cursor.getReadPosition(), markDeletedPosition.getNext());
     }
 
-    @Test(invocationCount = 5000)
+    @Test
     void testForceCursorRecovery() throws Exception {
         TestPulsarMockBookKeeper bk = new TestPulsarMockBookKeeper(executor);
         factory = new ManagedLedgerFactoryImpl(metadataStore, bk);

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
@@ -4873,7 +4873,7 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
                 final OpenCallback cb, final Object ctx) {
             if (ledgerErrors.containsKey(lId)) {
                 cb.openComplete(ledgerErrors.get(lId), null, ctx);
-            }else{
+            } else {
                 super.asyncOpenLedger(lId, digestType, passwd, cb, ctx);
             }
         }

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java
@@ -4840,24 +4840,17 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
         ManagedCursorInfo info = ManagedCursorInfo.newBuilder().setCursorsLedgerId(invalidLedger).build();
         CountDownLatch latch = new CountDownLatch(1);
         MutableBoolean recovered = new MutableBoolean(false);
-        AtomicBoolean callbackInvoked = new AtomicBoolean(false);
         VoidCallback callback = new VoidCallback() {
             @Override
             public void operationComplete() {
-                if (callbackInvoked.compareAndSet(false, true)) {
-                    log.info("Cursor recovery success");
-                    recovered.setValue(true);
-                    latch.countDown();
-                }
+                recovered.setValue(true);
+                latch.countDown();
             }
 
             @Override
             public void operationFailed(ManagedLedgerException exception) {
-                if (callbackInvoked.compareAndSet(false, true)) {
-                    log.error("Cursor recovery failed", exception);
-                    recovered.setValue(false);
-                    latch.countDown();
-                }
+                recovered.setValue(false);
+                latch.countDown();
             }
         };
         c1.recoverFromLedger(info, callback);
@@ -4880,9 +4873,9 @@ public class ManagedCursorTest extends MockedBookKeeperTestCase {
                 final OpenCallback cb, final Object ctx) {
             if (ledgerErrors.containsKey(lId)) {
                 cb.openComplete(ledgerErrors.get(lId), null, ctx);
-                return;
+            }else{
+                super.asyncOpenLedger(lId, digestType, passwd, cb, ctx);
             }
-            super.asyncOpenLedger(lId, digestType, passwd, cb, ctx);
         }
     }
 

